### PR TITLE
Remove leftover of type ascription feature gating

### LINKED ISSUE #111325: remove leftover `type_ascription` remains in feature gating
After #109128, there's some leftover code in feature gating.

https://github.com/rust-lang/rust/blob/613a5c95ae7401e08d21b8a77f53dadb1d7642d7/compiler/rustc_ast_passes/src/feature_gate.rs#L378-L401

It should be removed, although it needs to be ensured that code like it does not compile afterwards (I haven't tested it).

cc @chenyukang

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #111325

r? @Nilstrieb

### CODE REVIEW

File: compiler/rustc_ast_passes/src/feature_gate.rs [hunk 1]
```diff
@@ -376,43 +376,13 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
...
-            ).emit();
-        }
         visit::walk_stmt(self, stmt);
     }
 
     fn visit_expr(&mut self, e: &'a ast::Expr) {
         match e.kind {
             ast::ExprKind::Type(..) => {
-                if self.sess.parse_sess.span_diagnostic.err_count() == 0 {
-                    // To avoid noise about type ascription in common syntax errors,
```
Comment (Nilstrieb) [inline compiler/rustc_ast_passes/src/feature_gate.rs:407]: oh no

File: compiler/rustc_ast_passes/src/feature_gate.rs [hunk 2]
```diff
@@ -376,43 +376,13 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
...
-                    // only emit if it is the *only* error.
-                    gate_feature_post!(
-                        &self,
-                        type_ascription,
-                        e.span,
-                        "type ascription is experimental"
-                    );
-                } else {
+                if self.sess.parse_sess.span_diagnostic.err_count() > 0 {
```
Comment (Nilstrieb) [inline compiler/rustc_ast_passes/src/feature_gate.rs:385]: Can this be removed now? At least the comment below doesn't make sense without it.

File: compiler/rustc_ast_passes/src/feature_gate.rs [hunk 3]
```diff
@@ -376,54 +375,11 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
...
-                "type ascription is experimental",
-            ).span_suggestion_verbose(
-                lhs.span.shrink_to_lo(),
-                "you might have meant to introduce a new binding",
-                "let ",
-                Applicability::MachineApplicable,
-            ).emit();
-        }
         visit::walk_stmt(self, stmt);
     }
```
Comment (cjgillot) [inline compiler/rustc_ast_passes/src/feature_gate.rs:379]: The method override can be removed completely.

File: compiler/rustc_feature/src/active.rs [hunk 4]
```diff
@@ -532,8 +532,6 @@ declare_features! (
     (active, try_blocks, "1.29.0", Some(31436), None),
     /// Allows `impl Trait` to be used inside type aliases (RFC 2515).
     (active, type_alias_impl_trait, "1.38.0", Some(63063), None),
-    /// Allows the use of type ascription in expressions.
-    (active, type_ascription, "1.6.0", Some(23416), None),
```
[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_feature/src/active.rs:536]: Should the feature be declared in `rustc_feature::removed`?

[the remark this answers is not in the document]
Comment (Nilstrieb) [reply compiler/rustc_feature/src/active.rs:536]: I wonder how well that interacts with the library feature. Maybe we should just keep the feature gate, but you can try moving it to removed.

**Comment by Nilstrieb** [comment]
Let's keep the feature definition, but the removal of the gating code looks good.
r=me after reading the definition

**Pull request merged**